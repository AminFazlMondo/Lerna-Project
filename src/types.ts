import {javascript} from 'projen'

export interface LernaProjectOptions extends javascript.NodeProjectOptions {
  /**
   * (experimental) Use TypeScript for your projenrc file (`.projenrc.ts`).
   *
   * @default false
   * @experimental
   */
  readonly projenrcTs?: boolean;

  /**
   * Consolidate sub projects doc files
   *
   * @default false
   * @experimental
   */
  readonly docgen?: boolean;

  /**
   * (experimental) Docs directory.
   *
   * @default "docs"
   * @experimental
   */
  readonly docsDirectory?: string;

  /**
   * (experimental) Flag to run tasks only for the packages that has changes since last release
   *
   * @default false
   * @experimental
   */
  readonly sinceLastRelease?: boolean;

  /**
   * Whether or not to use Nx for task scheduling
   * https://lerna.js.org/docs/lerna-and-nx
   *
   * @default false
   */
  readonly useNx?: boolean;

  /**
   * Whether or not to use independent versioning for sub-projects
   * https://lerna.js.org/docs/features/version-and-publish#independent-mode
   *
   * @default false
   */
  readonly independentMode?: boolean;

}