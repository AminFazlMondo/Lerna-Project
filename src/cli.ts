import {Command} from 'commander'
import {copy, readdirSync, remove, existsSync, move} from 'fs-extra'

const program = new Command()

const distFolder = './dist'

program
  .command('clean-dist')
  .action(async () => {
    const entries = readdirSync(distFolder)
    await Promise.all(entries.map(f => remove(`${distFolder}/${f}`)))
  })

program
  .command('copy-dist <subProjectPath>')
  .action(async (subProjectPath) => {
    const subProjectDist = `./${subProjectPath}/dist`
    if (!existsSync(subProjectDist))
      return

    const entries = readdirSync(subProjectDist)
    await Promise.all(entries.map(f => copy(`${subProjectDist}/${f}`, `${distFolder}/${f}`, {recursive: true})))
  })


program
  .command('move-docs <parentDocsDirectory> <subProjectPath> <subProjectDocsDirectory>')
  .action(async (parentDocsDirectory, subProjectPath, subProjectDocsDirectory) => {
    await moveDocs(parentDocsDirectory, subProjectPath, subProjectDocsDirectory, false)
    await moveDocs(parentDocsDirectory, subProjectPath, 'API.md', true)
  })

async function moveDocs(parentDocsDirectory: string, subProjectPath: string, subPath: string, isFile: boolean): Promise<void> {
  const subProjectDocs = `./${subProjectPath}/${subPath}`
  if (!existsSync(subProjectDocs))
    return
  if (isFile) {
    await move(subProjectDocs, `./${parentDocsDirectory}/${subProjectPath}/${subPath}`)
  } else {
    const entries = readdirSync(subProjectDocs)
    await Promise.all(entries.map(f => move(`${subProjectDocs}/${f}`, `./${parentDocsDirectory}/${subProjectPath}/${f}`)))
  }
}

program.parse(process.argv)