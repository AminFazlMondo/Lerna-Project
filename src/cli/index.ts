export * from 'projen/lib/cli'