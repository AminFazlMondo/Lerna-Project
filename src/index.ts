import {JsonFile, javascript, Project, Tasks, SourceCode} from 'projen'
import {LernaProjectOptions} from './types'

export * from './types'

function getDocsDirectory(project: Project) {
  const result = Object.entries(project).find(([key]) => key === 'docsDirectory')
  return result?.[1].replace(/\/$/, '')
}

function getArtifactsDirectory(project: Project) {
  const result = Object.entries(project).find(([key]) => key === 'artifactsDirectory')
  return result?.[1]
}

const jsiiTaskPattern = /jsii-docgen -o (?<output>.+)$/i

function extractJsiiDocsOutput(tasks: Tasks): string | undefined {
  const jsiiDocStep = tasks.tryFind('docgen')?.steps.find(step => jsiiTaskPattern.test(step.exec ?? ''))
  if (!jsiiDocStep || !jsiiDocStep.exec)
    return

  const match = jsiiDocStep.exec.match(jsiiTaskPattern)
  return match?.groups?.output
}

const lockedTaskNames = ['build', 'upgrade', 'upgrade-projen']

export class LernaProject extends javascript.NodeProject {

  private subProjects: Record<string, Project>
  private projenrcTs: boolean

  readonly docsDirectory: string
  readonly docgen: boolean
  readonly sinceLastRelease: boolean
  readonly useNx: boolean
  readonly independentMode: boolean
  readonly useWorkspaces: boolean

  constructor(options: LernaProjectOptions) {
    super({
      ...options,
      jest: false,
    })

    this.addDevDeps('lerna-projen', 'lerna@5')

    if (options.projenrcTs)
      this.addDevDeps('ts-node', 'typescript')

    this.subProjects = {}
    this.docsDirectory = options.docsDirectory ?? 'docs'
    this.docgen = options.docgen ?? false
    this.sinceLastRelease = options.sinceLastRelease ?? false
    this.useNx = options.useNx ?? false
    this.projenrcTs = options.projenrcTs ?? false
    this.independentMode = options.independentMode ?? false
    this.useWorkspaces = options.useWorkspaces ?? false
  }

  addSubProject(subProject: Project) {
    const {outdir} = subProject

    if (!outdir || !outdir.includes(this.outdir))
      throw new Error('A sub project out dir should exists within the lerna package')

    const relativeOutDir = outdir.replace(`${this.outdir}/`, '')

    this.subProjects[relativeOutDir] = subProject
  }

  preSynthesize() {
    super.preSynthesize()
    const projenCommand = this.projenrcTs ? 'ts-node --skip-project .projenrc.ts' : 'node .projenrc.js'
    const {defaultTask} = this
    if (!defaultTask)
      throw new Error('Could not find default task')
    defaultTask.reset(projenCommand)
    this.packageTask.reset(`mkdir -p ${this.artifactsJavascriptDirectory}`)

    this.appendLernaCommands()

    this.preCompileTask.exec(`lerna-projen clean-dist ${this.artifactsDirectory}`)

    this.addCrossLinks()
    this.updateSubProjects()
    this.addDocumentsIndex()
  }

  private addCrossLinks() {
    const lernaConfig: any = {
      useNx: this.useNx,
      version: this.independentMode ? 'independent' : '0.0.0',
    }

    const packages = Object.keys(this.subProjects)

    if (this.useWorkspaces)
      this.package.addField('packages', packages)
    else
      lernaConfig.packages = packages


    new JsonFile(this, 'lerna.json', {
      obj: lernaConfig,
    })
  }

  private appendLernaCommands() {
    const upgradeTaskName = 'upgrade'
    const postUpgradeTaskName = 'post-upgrade'
    const postUpgradeTask = this.tasks.tryFind(postUpgradeTaskName)
    postUpgradeTask?.prependExec(this.getLernaCommand(upgradeTaskName, false))
    postUpgradeTask?.exec('npx projen')
    postUpgradeTask?.exec(this.getLernaCommand(postUpgradeTaskName))

    this.tasks.all
      .forEach(task => {
        if (lockedTaskNames.includes(task.name) || task.name === postUpgradeTaskName)
          return

        task.exec(this.getLernaCommand(task.name))
      })
  }

  private getLernaCommand(taskName: string, useSinceFlag = true) {
    const mainCommand = `lerna run ${taskName} --stream`
    const postCommand = useSinceFlag && this.sinceLastRelease ? ' --since $(git describe --abbrev=0 --tags --match "v*")' : ''
    return `${mainCommand}${postCommand}`
  }

  private updateSubProjects() {
    const bumpTask = this.tasks.tryFind('bump')
    const unbumpTask = this.tasks.tryFind('unbump')

    Object.entries(this.subProjects).forEach(([subProjectPath, subProject]) => {
      const subProjectDocsDirectory = getDocsDirectory(subProject)
      if (this.docgen && subProjectDocsDirectory)
        this.postCompileTask.exec(`lerna-projen move-docs ${this.docsDirectory} ${subProjectPath} ${subProjectDocsDirectory}`)

      const packageAllTask = subProject.tasks.tryFind('package-all')

      if (packageAllTask)
        subProject.packageTask.spawn(packageAllTask)

      const artifactsDirectory = getArtifactsDirectory(subProject)

      this.packageTask.exec(`lerna-projen copy-dist ${subProjectPath}/${artifactsDirectory} ${this.artifactsDirectory}`)

      subProject.defaultTask?.reset()

      const bumpEnvs = {
        OUTFILE: 'package.json',
        CHANGELOG: `${artifactsDirectory}/changelog.md`,
        BUMPFILE: `${artifactsDirectory}/version.txt`,
        RELEASETAG: `${artifactsDirectory}/releasetag.txt`,
        RELEASE_TAG_PREFIX: '',
      }

      if (bumpTask) {
        const subBumpTask = subProject.addTask(bumpTask.name, {
          description: bumpTask.description,
          condition: bumpTask.condition,
          env: bumpEnvs,
        })
        subBumpTask.builtin('release/bump-version')
      }

      if (unbumpTask) {
        const subBumpTask = subProject.addTask(unbumpTask.name, {
          description: unbumpTask.description,
          condition: unbumpTask.condition,
          env: bumpEnvs,
        })
        subBumpTask.builtin('release/reset-version')
      }
    })
  }

  private addDocumentsIndex() {
    if (!this.docgen)
      return

    const subProjectsDocs: Record<string, string> = {}

    const indexMarkdown = new SourceCode(this, `${this.docsDirectory}/index.md`)
    const readmeMarkdown = new SourceCode(this, `${this.docsDirectory}/README.md`)

    Object.entries(this.subProjects).forEach(([subProjectPath, subProject]) => {
      const subProjectDocsDirectory = getDocsDirectory(subProject)
      if (!subProjectDocsDirectory)
        return

      const jsiiDocsOutput = extractJsiiDocsOutput(subProject.tasks)
      if (jsiiDocsOutput) {
        const path = `${subProjectPath}/${jsiiDocsOutput}`
        indexMarkdown.line(`- ## [${subProject.name}](${path})`)
        subProjectsDocs[subProject.name] = path
      } else {
        subProjectsDocs[subProject.name] = `${subProjectPath}/index.html`
      }
      readmeMarkdown.line(`- ## [${subProject.name}](../${subProjectPath}/README.md)`)
    })

    const indexHtml = new SourceCode(this, `${this.docsDirectory}/index.html`)
    indexHtml.line('<!DOCTYPE html>')
    indexHtml.line('<html>')
    indexHtml.open('<body>')
    indexHtml.open('<ul>')
    Object.entries(subProjectsDocs).forEach(([name, path]) => {
      indexHtml.open('<li>')
      indexHtml.open(`<a href="${path}">`)
      indexHtml.line(name)
      indexHtml.close('</a>')
      indexHtml.close('</li>')
    })
    indexHtml.close('</ul>')
    indexHtml.close('</body>')
    indexHtml.line('</html>')
    indexHtml.line()

    this.gitattributes.addAttributes(`/${this.docsDirectory}/**`, 'linguist-generated')
  }
}