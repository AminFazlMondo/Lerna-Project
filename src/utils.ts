import {NodeProject} from 'projen/lib/javascript'

/**
 * Adds dependency to the project for NX task runner
 * see https://nx.dev/reference/project-configuration#dependson
 *
 * @param project The Project to add dependency to
 * @param taskName The task name that is dependent on another tasks
 * @param dependsOnTaskName The task name that is dependent on in other projects
 * @param dependsOn The packages that source project is dependent on
 */
export function addNxTaskDependency(project: NodeProject, taskName: string, dependsOnTaskName: string, ...dependsOn: NodeProject[]): void {
  project.package.addField('nx', {
    targets: {
      [taskName]: {
        dependsOn: [
          {
            projects: dependsOn.map(d => d.package.packageName),
            target: dependsOnTaskName,
          },
        ],
      },
    },
  })
}