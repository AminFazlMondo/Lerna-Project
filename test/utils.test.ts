import {NodeProject} from 'projen/lib/javascript'
import {synthSnapshot} from 'projen/lib/util/synth'
import {addNxTaskDependency} from '../src/utils'

describe('utils', () => {
  describe('addNxTaskDependency', () => {
    test('should add dependency in nx', () => {
      const project1 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-1',
      })
      const project2 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-2',
      })
      const project3 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-3',
      })

      addNxTaskDependency(project1, 'test', 'lint', project2, project3)
      addNxTaskDependency(project2, 'compile', 'pre-compile', project1)

      const output1 = synthSnapshot(project1)
      const output2 = synthSnapshot(project2)

      expect(output1['package.json'].nx).toMatchSnapshot()
      expect(output2['package.json'].nx).toMatchSnapshot()
    })
  })
})