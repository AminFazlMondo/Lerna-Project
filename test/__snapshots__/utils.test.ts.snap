// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils addNxDependency deprecated should add dependency in nx when both task and project dependency is specified 1`] = `
{
  "implicitDependencies": [
    "project-2",
  ],
  "targets": {
    "test": {
      "dependsOn": [
        {
          "projects": [
            "project-3",
          ],
          "target": "lint",
        },
      ],
    },
  },
}
`;

exports[`utils addNxDependency deprecated should add dependency in nx when only task dependency is specified 1`] = `
{
  "targets": {
    "test": {
      "dependsOn": [
        {
          "projects": [
            "project-2",
            "project-3",
          ],
          "target": "lint",
        },
      ],
    },
  },
}
`;

exports[`utils addNxDependency should add cache flag when specified 1`] = `
{
  "targets": {
    "default": {
      "dependsOn": [
        {
          "projects": [
            "project-2",
          ],
          "target": "build",
        },
      ],
    },
    "test": {
      "dependsOn": [
        {
          "cache": true,
          "projects": [
            "project-3",
          ],
          "target": "lint",
        },
      ],
    },
  },
}
`;

exports[`utils addNxDependency should add dependency in nx when both task and project dependency is specified 1`] = `
{
  "implicitDependencies": [
    "project-2",
  ],
  "targets": {
    "default": {
      "dependsOn": [
        {
          "projects": [
            "project-3",
          ],
          "target": "build",
        },
      ],
    },
    "test": {
      "dependsOn": [
        {
          "projects": [
            "project-3",
          ],
          "target": "lint",
        },
      ],
    },
  },
}
`;

exports[`utils addNxDependency should add dependency in nx when only project dependency is specified 1`] = `
{
  "implicitDependencies": [
    "project-2",
    "project-3",
  ],
}
`;

exports[`utils addNxDependency should add dependency in nx when only task dependency is specified 1`] = `
{
  "targets": {
    "test": {
      "dependsOn": [
        {
          "projects": [
            "project-2",
            "project-3",
          ],
          "target": "lint",
        },
      ],
    },
  },
}
`;

exports[`utils addNxProjectDependency should add dependency in nx 1`] = `
{
  "implicitDependencies": [
    "project-2",
    "project-3",
  ],
}
`;

exports[`utils addNxProjectDependency should add dependency in nx 2`] = `
{
  "implicitDependencies": [
    "project-3",
  ],
}
`;

exports[`utils addNxTaskDependency should add dependency in nx 1`] = `
{
  "targets": {
    "test": {
      "dependsOn": [
        {
          "projects": [
            "project-2",
            "project-3",
          ],
          "target": "lint",
        },
      ],
    },
  },
}
`;

exports[`utils addNxTaskDependency should add dependency in nx 2`] = `
{
  "targets": {
    "compile": {
      "dependsOn": [
        {
          "projects": [
            "project-1",
          ],
          "target": "pre-compile",
        },
      ],
    },
  },
}
`;
